storage/conformance_test.go

```diff
@@ -16,21 +16,337 @@ package storage
+						t.Logf("No tests for operation %v", method.Name)
+					}
+					for i, fn := range methods[method.Name] {
+						testName := fmt.Sprintf("%v-%v-%v-%v", retryTest.Id, instructions.Instructions, method.Name, i)
+						t.Run(testName, func(t *testing.T) {
+
+							// Create the retry subtest
+							subtest := &retrySubtest{T: t, name: testName}
+							err := subtest.create(host, map[string][]string{
+								method.Name: instructions.Instructions,
+							})
+							if err != nil {
+								t.Fatalf("setting up retry test: %v", err)
+							}
+
+							// Create necessary test resources in the emulator
+							fs := &resources{}
+							for _, resource := range method.Resources {
+								if err := fs.populate(ctx, client, resource); err != nil {
+									t.Fatalf("creating test resources: %v", err)
+								}
+							}
+
+							// Test
+							err = fn(ctx, subtest.wrappedClient, fs, retryTest.PreconditionProvided)
+							if retryTest.ExpectSuccess && err != nil {
+								t.Errorf("want success, got %v", err)
+							}
+							if !retryTest.ExpectSuccess && err == nil {
+								t.Errorf("want failure, got success")
+							}
+
+							// Verify that all instructions were used up during the test
+							// (indicates that the client sent the correct requests).
+							if err := subtest.check(); err != nil {
+								t.Errorf("checking instructions: %v", err)
+							}
+
+							// Close out test in emulator.
+							if err := subtest.delete(); err != nil {
+								t.Errorf("deleting retry test: %v", err)
+							}
+						})
+					}
+
+				}
+			}
+		}
+	}
+
+}
+
+type retrySubtest struct {
+	*testing.T
+	name          string
+	id            string   // ID to pass as a header in the test execution
+	host          *url.URL // set the path when using; path is not guaranteed betwen calls
+	wrappedClient *Client
+}
+
+// Create a retry test resource in the emulator
+func (rt *retrySubtest) create(host string, instructions map[string][]string) error {
+	endpoint, err := parseURL(host)
+	if err != nil {
+		return err
+	}
+	rt.host = endpoint
+
+	c := http.DefaultClient
+	data := struct {
+		Instructions map[string][]string `json:"instructions"`
+	}{
+		Instructions: instructions,
+	}
+
+	buf := new(bytes.Buffer)
+	if err := json.NewEncoder(buf).Encode(data); err != nil {
+		return fmt.Errorf("encoding request: %v", err)
+	}
+
+	rt.host.Path = "retry_test"
+	resp, err := c.Post(rt.host.String(), "application/json", buf)
+	if err != nil || resp.StatusCode != 200 {
+		return fmt.Errorf("creating retry test: err: %v, resp: %+v", err, resp)
+	}
+	defer func() {
+		closeErr := resp.Body.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+	testRes := struct {
+		TestID string `json:"id"`
+	}{}
+	if err := json.NewDecoder(resp.Body).Decode(&testRes); err != nil {
+		return fmt.Errorf("decoding test ID: %v", err)
+	}
+
+	rt.id = testRes.TestID
+
+	// Create wrapped client which will send emulator instructions
+	rt.host.Path = ""
```

Comment:
Looks good! Curious is the main purpose of creating a wrappedClient in here to populate the retrySubtest struct?